doc/fixed-point-arguments: restructure chapter to make it more approachable (#537509)
NixOS/nixpkgs: doc/build-helpers/fixed-point-arguments.chapter.md

# Fixed-point arguments of build helpers {#chap-build-helpers-finalAttrs}

`stdenv.mkDerivation` also accepts a [fixed-point function](#function-library-lib.fixedPoints.fix) instead of a plain attribute set:

```nix
stdenv.mkDerivation (finalAttrs: {
  pname = "hello";
  version = "2.12";

  src = fetchurl {
    url = "mirror://gnu/hello/hello-${finalAttrs.version}.tar.gz";
    hash = "sha256-...";
  };
})
```

The function's input, conventionally named `finalAttrs`, is the final state of the attribute set. Here `src` reads `finalAttrs.version` instead of repeating the version string. A build helper like this is said to accept **fixed-point arguments**.

Attributes that reference each other through `finalAttrs` stay correct when changing any of them with [`overrideAttrs`](#sec-pkg-overrideAttrs), because they all access the final values of the fixed-point computation.

`rec` cannot do this: its self-references are fixed when the set is defined and ignore later overrides.
See [recursive-sets](https://nix.dev/manual/nix/stable/language/syntax#recursive-sets) for the underlying mechanism.

## Define a build helper with `lib.extendMkDerivation` {#sec-build-helper-extendMkDerivation}

> Use [`lib.customisation.extendMkDerivation`](#function-library-lib.customisation.extendMkDerivation) to define a build helper with fixed-point support from an existing one. It takes an attribute overlay similar to [`<pkg>.overrideAttrs`](#sec-pkg-overrideAttrs).

Reviewer: Should we also mention that such attribute overlay should be specified with `extendDrvArgs`, or is it obvious enough in the example?

Author: Should definitely mention currently reads as if overrideAttrs would be the mechanism for downstream overrides. 

Can you open another PR ideally just add some examples with little explanation. 

Order Example > Explanation

> Besides overriding, `lib.extendMkDerivation` also supports `excludeDrvArgNames` to optionally exclude some arguments in the input fixed-point arguments from passing down to the base build helper (specified as `constructDrv`).

Reviewer: I find it clever that you wrote about excluding "some arguments in the input fixed-point arguments"!
Those arguments can still be passed down, but only if `extendDrvArgs` specifies them.

# Fixed-point arguments of build helpers {#chap-build-helpers-finalAttrs}

> `stdenv.mkDerivation` accepts a fixed-point function instead of a plain attribute set:

Reviewer 2: ```suggestion
`stdenv.mkDerivation` also accepts a fixed-point function instead of a plain attribute set:
```

Reviewer 2: Also may be worth linking "fixed-point function" to `lib.fix` docs. Fixed points are one of the most magical concepts in Nixpkgs, and the earlier new readers find that piece, which IMO does a very good job at demystifying it, the better.

```nix
stdenv.mkDerivation (finalAttrs: {
  pname = "hello";
  version = "2.12";

  src = fetchurl {
    url = "mirror://gnu/hello/hello-${finalAttrs.version}.tar.gz";
    hash = "sha256-...";
  };
})
```

The function's input, conventionally named `finalAttrs`, is the final state of the attribute set. Here `src` reads `finalAttrs.version` instead of repeating the version string. A build helper like this is said to accept **fixed-point arguments**.

> This is what makes a package easy to override: attributes that reference each other through `finalAttrs` stay correct when someone changes one of them, because they all resolve to the final values. For an example of what an override looks like, see [`<pkg>.overrideAttrs`](#sec-pkg-overrideAttrs).

Reviewer 2: Never use value judgement. Something easy for you may be insurmountable for someone else. It is sometimes okay to compare how easy one thing is versus another, if you can make it evident through demonstration. But here it doesn't even make much sense, because you'd be advertising instead of listing what's there: "this is easier than that" -- so why do we have the less easy thing in the first place? That's for explanation (why things were made a certain way, could be done e.g. in the release notes) to answer, not reference documentation.

Just state the facts.

```suggestion
Attributes that reference each other through `finalAttrs` stay correct when changing any of them with [`overrideAttrs`](#sec-pkg-overrideAttrs), because they all access the final values of the fixed-point computation.
```

Author: Right, i might add this to the styleguide. Not sure if I just forgot about it and it's already there

Reviewer 2: Seems like I never wrote that down... Sure worth adding!

`rec` cannot do this: its self-references are fixed when the set is defined and ignore later overrides.

> See [recursive attributes in `mkDerivation`](#mkderivation-recursive-attributes) for the underlying mechanism.

Reviewer 2: Ambiguous reference. Despite the link text this may as well refer to the underlying mechanism of `rec`.

> Support for fixed-point arguments was added to [`stdenv.mkDerivation`](#mkderivation-recursive-attributes) in Nixpkgs 22.05.

Reviewer 2: Not sure that's needed; doesn't this belong to the release notes? And this repeats the same link as in the sentence before.

I understand that sometimes it's nice to know when some piece of API was added, but that should ideally be structured data and not some ad hoc sentence somewhere. Which is why I'd say this particular thing is very out of scope for this change.

Author: I'll remove it. People can retrieve this from release notes. We might want some a feature compat table at some point. Not sure to what depth yet.

Reviewer 2: Definitely a rabbit-hole for another side quest... :)

```
:::

> If you need to apply extra changes to the result derivation, pass the derivation transformation function to `lib.extendMkDerivation`:

Reviewer 2: ```suggestion
To apply extra changes to the result derivation, pass the derivation transformation function to `lib.extendMkDerivation`:
```
